reach-whitepaper.md

```diff
@@ -0,0 +1,1019 @@
+constraint, the scale factor is now divided by the number of levels in addition
+to dividing it by the contribution cap (`CONTRIBUTION_CAP`). 
+While we adhere to equal contribution per level, it is possible to further
+optimize the contribution budget across different levels, as explained in the
+[relevant paper](https://arxiv.org/pdf/2308.13510).  
+
+The code for this is below.
+
+
+```javascript
+const CONTRIBUTION_CAP = 5;
+const NUM_LEVELS = 4;
+const L1_BUDGET = 65536;
+
+function convertContentIdAndLevelToBucket(contentId, level, ...otherDimensions) {
+  // Implement a function that computes the bucket key for a given level using 
+  // level-related fields from otherDimensions.
+  // E.g., in our example, the second level could set the key to be the concatenation
+  // between advertiser and campaign id of this content.
+}
+class ReachMeasurementOperation {
+  async run(data) {
+    const { contentId, ...otherDimensions } = data;
+
+    for (let level = 1; level <= NUM_LEVELS; level++) {
+      // Generate the aggregation key and the aggregatable value
+      const bucket = convertContentIdAndLevelToBucket(contentId, level, ...otherDimensions);
+      const value = Math.floor(L1_BUDGET / (CONTRIBUTION_CAP * NUM_LEVELS));
+
+      // Read from Shared Storage
+      const key = `has-reported-content-${bucket}`;
+      const hasReportedContent = (await sharedStorage.get(key)) === 'true';
+
+      // Do not report if a report has been sent already
+      if (hasReportedContent) {
+        return;
+      }
+
+      // Send an aggregatable report via the Private Aggregation API
+      privateAggregation.contributeToHistogram({ bucket, value });
+
+      // Set the report submission status flag
+      await sharedStorage.set(key, true);
+    }
+  }
+}
+
+// Register the operation
+register('reach-measurement', ReachMeasurementOperation);
+```
+
+
+A sample call would look like:
+
+
+```javascript
+await window.sharedStorage.run('reach-measurement', { data: {
+    contentId: '123', 
+    geo: 'san jose', 
+    creativeId: '55' 
+}});
+```
+
+
+
+### Cumulative Reach Measurements
+
+So far, we have not mentioned the time aspect of the queries. 
+In particular, we assume that all queries correspond to the same time periods. 
+However, we may want to measure time-related queries. 
+A popular family of such queries is cumulative queries, which are queries of 
+the form “what is the total reach up until timestep `t`?” where 
+`t` varies from `1, …, T`.
+For example, one might want to measure the reach up until today everyday for the
+ entire month. In this case, we would have `T = 30` and “timestep” being a day.
+
+_Note: Both methods described in this section are limited by the 
+[retention policy](https://github.com/WICG/shared-storage?tab=readme-ov-file#data-retention-policy) of Shared Storage which is 30 days from last write._
```

Comment:
hyper nit: could remove the `?tab=readme-ov-file`

Reply:
Done.